### PR TITLE
fix: check if a key is defined before dereferencing it

### PR DESCRIPTION
**Pull request by Zhicheng-Liu**
Due to Perl's autovivification feature, if a key is not defined, dereferencing
it will automaticaly create an empty hash as value for that key. This will be
an issue when checking if the key is defined elsewhere.

This change remove this potential problem for 'DATABASE_FUNCGEN' and 'DATABASE
_VARIATION' keys in species_defs->databases.

Related to ENSEMBL-4611.

### CODE REVIEW

**Comment by andrewyatz** [comment]
I'm slightly worried that you won't end up with massive autoviv happening but there will still be some. Might be better to change statements such as 

```perl
if($self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}){
```

To

```perl
if( exists $self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}) {
```

That should stop any autoviv happening and ensures you're testing the existence of the key

**Comment by Zhicheng-Liu** [comment]
@andrewyatz I think checking existence only will also let an existed key that has undef value pass through, e.g. when $self->hub->species_defs->databases->{'DATABASE_FUNCGEN'} = undef. By making sure that this key does have associated value will ignore  such cases and it is consistent with testing conditions found elsewhere.

BTW doing this (if($self->hub->species_defs->databases->{'DATABASE_FUNCGEN'})) will not invoke autovivification. Only dereferencing it will.

**Comment by andrewyatz** [comment]
That's fair enough 👍

File: modules/EnsEMBL/Web/Component/Gene/RegulationImage.pm [hunk 1]
```diff
@@ -46,7 +46,9 @@ sub content {
   my $key = $wuc->get_track_key( 'transcript', $object );
   $wuc->modify_configs( [$key], {qw(display collapsed_label)} );
 
-  $wuc->{'data_by_cell_line'} = $self->new_object('Slice', $extended_slice, $object->__data)->get_cell_line_data($wuc) if keys %{$self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'cell_type'}{'ids'}};
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Component/Gene/RegulationImage.pm]: This one is ok.

File: modules/EnsEMBL/Web/Component/Gene/RegulationImage.pm [hunk 2]
```diff
@@ -46,7 +46,9 @@ sub content {
...
+  if ( $self->hub->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    $wuc->{'data_by_cell_line'} = $self->new_object('Slice', $extended_slice, $object->__data)->get_cell_line_data($wuc) if keys %{$self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'cell_type'}{'ids'}};
+  }
 
   my $image    = $self->new_image( $extended_slice, $wuc, [] );
   $image->imagemap           = 'yes';
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Component/Gene/RegulationImage.pm]: This part is ok.

File: modules/EnsEMBL/Web/Component/Gene/TranscriptsImage.pm [hunk 3]
```diff
@@ -56,7 +56,9 @@ sub content {
   
   $node->set('display', 'transcript_label') if $node && $node->get('display') eq 'off';
 
-  $image_config->{'data_by_cell_line'} = $self->new_object('Slice', $gene_slice, $object->__data)->get_cell_line_data_closure($image_config) if keys %{$self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'cell_type'}{'ids'}};
+  if ( $self->hub->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    $image_config->{'data_by_cell_line'} = $self->new_object('Slice', $gene_slice, $object->__data)->get_cell_line_data_closure($image_config) if keys %{$self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'cell_type'}{'ids'}};
+  }
 
   my $image = $self->new_image($gene_slice, $image_config, [ $gene->stable_id ]);
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Component/Gene/TranscriptsImage.pm]: This part is ok.

File: modules/EnsEMBL/Web/Component/Location/ViewBottom.pm [hunk 4]
```diff
@@ -57,11 +57,13 @@ sub content {
...
   # Add multicell configuration
-  $image_config->{'data_by_cell_line'} = $self->new_object('Slice', $slice, $object->__data)->get_cell_line_data_closure($image_config) if keys %{$hub->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'cell_type'}{'ids'}};
+  if ( $hub->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    $image_config->{'data_by_cell_line'} = $self->new_object('Slice', $slice, $object->__data)->get_cell_line_data_closure($image_config) if keys %{$hub->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'cell_type'}{'ids'}};
+  }
   $image_config->_update_missing($object);
-  
+
   my $info  = $self->_add_object_track($image_config);
   my $image = $self->new_image($slice, $image_config, $object->highlights);
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Component/Location/ViewBottom.pm]: This part is ok.

File: modules/EnsEMBL/Web/Component/Regulation/Summary.pm [hunk 5]
```diff
@@ -72,7 +72,10 @@ sub content {
...
   my @class = ($object->feature_type->name);
 
-  my $epigenome_count = grep { $_ > 0 } values %{$self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}->{'tables'}{'cell_type'}{'ids'}}; 
+  my $epigenome_count = 0;
+  if ( $self->hub->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    $epigenome_count = grep { $_ > 0 } values %{$self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}->{'tables'}{'cell_type'}{'ids'}};
+  }
 
   $summary->add_row('Classification',join(', ',@class));
   $summary->add_row('Location', $location_html);
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Component/Regulation/Summary.pm]: This part is ok.

File: modules/EnsEMBL/Web/Component/Transcript/ProteinVariations.pm [hunk 6]
```diff
@@ -101,7 +101,8 @@ sub content {
   ];
  
   # add SIFT for supported species
-  if ( $hub->species_defs->databases->{'DATABASE_VARIATION'}->{'SIFT'}){
+  if ( $hub->species_defs->databases->{'DATABASE_VARIATION'} &&
+       $hub->species_defs->databases->{'DATABASE_VARIATION'}->{'SIFT'}){
     push @$columns, ({ key => 'sift', title => 'SIFT', width => '8%', align => 'center', sort => 'position_html', help => $glossary->{'SIFT'} });
   }
   if ($hub->species =~ /homo_sapiens/i) {
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Component/Transcript/ProteinVariations.pm]: This part is ok.

File: modules/EnsEMBL/Web/Component/Variation/HighLD.pm [hunk 7]
```diff
@@ -44,7 +44,7 @@ sub content {
   my $species_defs = $hub->species_defs;
   
   # check we have a location and LD populations are defined
-  return unless $self->builder->object('Location') && $species_defs->databases->{'DATABASE_VARIATION'}{'DEFAULT_LD_POP'};
+  return unless $self->builder->object('Location') && $species_defs->databases->{'DATABASE_VARIATION'} && $species_defs->databases->{'DATABASE_VARIATION'}{'DEFAULT_LD_POP'};
   
   my $selected_pop = $hub->param('pop_id');
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Component/Variation/HighLD.pm]: This part is ok.

File: modules/EnsEMBL/Web/Component/Variation/Mappings.pm [hunk 8]
```diff
@@ -433,7 +433,7 @@ sub table_columns {
   );  
 
   push @columns, ({ key => 'sift',     title => 'SIFT',     sort => 'position_html', align => 'center', help => $glossary->{'SIFT'} })
-      if defined $self->hub->species_defs->databases->{'DATABASE_VARIATION'}->{'SIFT'};
+      if $self->hub->species_defs->databases->{'DATABASE_VARIATION'} && defined $self->hub->species_defs->databases->{'DATABASE_VARIATION'}->{'SIFT'};
 
   push @columns, ({ key => 'polyphen', title => 'PolyPhen', sort => 'position_html', align => 'center', help => $glossary->{'PolyPhen'} })
       if $self->hub->species eq 'Homo_sapiens';
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Component/Variation/Mappings.pm]: This part is ok.

File: modules/EnsEMBL/Web/DBSQL/DBConnection.pm [hunk 9]
```diff
@@ -135,7 +135,9 @@ sub get_DBAdaptor {
   # warn "$species - $database - $dba";
 
   # Funcgen Database Files Overwrite
-  if ($database eq 'funcgen' && $self->{'species_defs'}->databases->{'DATABASE_FUNCGEN'}{'NAME'}) {
+  if ($database eq 'funcgen' &&
+      $self->{'species_defs'}->databases->{'DATABASE_FUNCGEN'} &&
+      $self->{'species_defs'}->databases->{'DATABASE_FUNCGEN'}{'NAME'}) {
     my $file_path = join '/', $self->{'species_defs'}->DATAFILE_BASE_PATH, lc $species, $self->{'species_defs'}->ASSEMBLY_VERSION;
     $dba->get_ResultSetAdaptor->dbfile_data_root($file_path) if ($dba && -e $file_path && -d $file_path);
   }
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/DBSQL/DBConnection.pm]: This part is ok.

File: modules/EnsEMBL/Web/Factory/StructuralVariation.pm [hunk 10]
```diff
@@ -61,7 +61,7 @@ sub createObjects {
     $self->param('vdb', 'variation');
     $self->param('sv', $structural_variation->variation_name) unless $self->param('sv'); # For same reason as svf check above;
   } else {
-    my $db_version = $self->species_defs->databases->{'DATABASE_VARIATION'}{'dbSNP_VERSION'};
+    my $db_version = $self->species_defs->databases->{'DATABASE_VARIATION'} && $self->species_defs->databases->{'DATABASE_VARIATION'}{'dbSNP_VERSION'};
     
     return $self->problem('fatal', "Could not find structural variation $identifier", $self->_help(sprintf(
       'Either %s does not exist in the current Ensembl database, %sor there was a problem retrieving it.',
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Factory/StructuralVariation.pm]: This part is ok.

File: modules/EnsEMBL/Web/ImageConfig/contigviewbottom.pm [hunk 11]
```diff
@@ -314,7 +314,8 @@ sub init_cacheable {
   }
   elsif ($self->hub->database('variation')) {
     my $tracks = [qw(variation_feature_variation)];
-    if ($self->species_defs->databases->{'DATABASE_VARIATION'}{'STRUCTURAL_VARIANT_COUNT'}) {
+    if ($self->species_defs->databases->{'DATABASE_VARIATION'} &&
+        $self->species_defs->databases->{'DATABASE_VARIATION'}{'STRUCTURAL_VARIANT_COUNT'}) {
       push @$tracks, 'variation_feature_structural_smaller';
     }
     $self->modify_configs($tracks, {display => 'compact'});
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/ImageConfig/contigviewbottom.pm]: This part is ok.

File: modules/EnsEMBL/Web/ImageConfig/regulation_view.pm [hunk 12]
```diff
@@ -36,7 +36,10 @@ sub init_cacheable {
...
 
   my @feature_sets  = ('cisRED', 'VISTA', 'miRanda', 'NestedMICA', 'REDfly CRM', 'REDfly TFBS', 'search');
-  my $cell_info     = $self->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'cell_type'}{'ids'};
+  my $cell_info     = {};
+  if ( $self->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    $cell_info      = $self->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'cell_type'}{'ids'};
+  }
   my @cell_lines    = grep { $cell_info->{$_} > 0 } sort keys %$cell_info;
 
   s/\:\d*$// for @cell_lines;
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/ImageConfig/regulation_view.pm]: This part is ok.

File: modules/EnsEMBL/Web/ImageConfigExtension/Tracks.pm [hunk 13]
```diff
@@ -1830,7 +1836,7 @@ sub add_somatic_mutations {
     my $tissue_menu = $self->create_menu_node('somatic_mutation_by_tissue', 'Somatic variants by tissue');
 
     ## Add tracks for each tumour site
-    my %tumour_sites = %{$self->species_defs->databases->{'DATABASE_VARIATION'}{'SOMATIC_MUTATIONS'}{$key_2} || {}};
+    my %tumour_sites = ($self->species_defs->databases->{'DATABASE_VARIATION'} && %{$self->species_defs->databases->{'DATABASE_VARIATION'}{'SOMATIC_MUTATIONS'}{$key_2} || {}}) || {};
 
     foreach my $description (sort  keys %tumour_sites) {
       next if $description eq 'none';
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/ImageConfigExtension/Tracks.pm]: This part is ok.

File: modules/EnsEMBL/Web/Object/Gene.pm [hunk 14]
```diff
@@ -1622,7 +1622,10 @@ sub get_extended_reg_region_slice {
...
   my $self = shift;
 
-  my $available_sets = $self->species_defs->databases->{'DATABASE_FUNCGEN'}->{'FEATURE_SETS'};
+  my $available_sets = [];
+  if ( $self->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    $available_sets = $self->species_defs->databases->{'DATABASE_FUNCGEN'}->{'FEATURE_SETS'};
+  }
   my $fg_db = $self->get_fg_db; 
   my $feature_set_adaptor = $fg_db->get_FeatureSetAdaptor;
   my @fsets;
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Object/Gene.pm]: This part is ok.

File: modules/EnsEMBL/Web/Object/LRG.pm [hunk 15]
```diff
@@ -1036,7 +1036,10 @@ return $fg_db;
...
   my $self = shift;
 
-  my $available_sets = $self->species_defs->databases->{'DATABASE_FUNCGEN'}->{'FEATURE_SETS'};
+  my $available_sets = [];
+  if ( $self->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    $available_sets = $self->species_defs->databases->{'DATABASE_FUNCGEN'}->{'FEATURE_SETS'};
+  }
   my $fg_db = $self->get_fg_db; 
   my $feature_set_adaptor = $fg_db->get_FeatureSetAdaptor;
   my @fsets;
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Object/LRG.pm]: This part is ok.

File: modules/EnsEMBL/Web/Object/Regulation.pm [hunk 16]
```diff
@@ -335,12 +335,22 @@ sub get_evidence_data {
...
-  return [sort keys %{$self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}->{'tables'}{'cell_type'}{'regbuild_names'}}];
+
+  if ( $self->hub->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    return [sort keys %{$self->hub->species_defs->databases->{'DATABASE_FUNCGEN'}->{'tables'}{'cell_type'}{'regbuild_names'}}];
+  }
+
+  return [];
 }
 
 ################ Calls for Feature in Detail view ###########################
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Object/Regulation.pm]: This part is ok.

File: modules/EnsEMBL/Web/Object/Slice.pm [hunk 17]
```diff
@@ -346,7 +346,10 @@ sub get_cell_line_data {
   my ($self, $image_config, $filter) = @_;
   
   # First work out which tracks have been turned on in image_config
-  my %cell_lines = %{$self->species_defs->databases->{'DATABASE_FUNCGEN'}->{'tables'}{'cell_type'}{'ids'}};
+  my %cell_lines = ();
+  if ( $self->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    %cell_lines = %{$self->species_defs->databases->{'DATABASE_FUNCGEN'}->{'tables'}{'cell_type'}{'ids'}};
+  }
   my @sets       = qw(core non_core);  
   my $data;
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Object/Slice.pm]: This part is ok,

File: modules/EnsEMBL/Web/Object/Variation.pm [hunk 18]
```diff
@@ -1674,7 +1674,7 @@ sub hgvs_url {
       $p->{'t'}      = $refseq.$version;
     } else { # $type eq c: cDNA position, no $type: special cases where the variation falls in e.g. a pseudogene. Default to transcript
       $p->{'type'}   = 'Transcript';
-      $p->{'action'} = ($hub->species_defs->databases->{'DATABASE_VARIATION'}->{'#STRAINS'} > 0 ? 'Population' : 'Summary');
+      $p->{'action'} = ($hub->species_defs->databases->{'DATABASE_VARIATION'} && $hub->species_defs->databases->{'DATABASE_VARIATION'}->{'#STRAINS'} > 0 ? 'Population' : 'Summary');
       $p->{'t'}      = $refseq.$version;
     }
   }
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/Object/Variation.pm]: This part is ok.

File: modules/EnsEMBL/Web/ViewConfig/Regulation/FeatureDetails.pm [hunk 19]
```diff
@@ -27,7 +27,10 @@ use parent qw(EnsEMBL::Web::ViewConfig::RegulationPage);
...
   ## Abstract method implementation
   my $self     = shift;
-  my $analyses = $self->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'feature_type'}{'analyses'} || {};
+  my $analyses = {};
+  if ( $self->species_defs->databases->{'DATABASE_FUNCGEN'} ) {
+    $analyses = $self->species_defs->databases->{'DATABASE_FUNCGEN'}{'tables'}{'feature_type'}{'analyses'} || {};
+  }
 
   $self->set_default_options({
     'context'   => 200,
```
Comment (ens-ds23) [inline modules/EnsEMBL/Web/ViewConfig/Regulation/FeatureDetails.pm]: This part is ok.

**Pull request merged**